### PR TITLE
CRe, MuPDF, PicDocument: add support for WebP images

### LINKED ISSUE #3016: WebP images support on CBZ files
* KOReader version: v2015.11-3-g9e0be45  
* Device: Kindle Paperwhite 3  

#### Issue  
When I try to open a CBZ with WebP files inside, file manager return to list of files.    

#### Steps to reproduce  
Put some of WebP files numbered in a zip, rename the file to "cbz" extension, put in Kindle storage, launch KOreader, open file manager, navigate to folder with the cbz file, try to open.  
  
Note: some of scanlators now is making the mangas with WebP format for fast and low bandwidth reading in browser, I have to convert every manga to jpg and repack in a zip for reading in KOreader.

### LINKED ISSUE #9240: in book images not appearing properly
* KOReader version: KOReader 2022.05.1
* Device: poco x3 pro 256/8 gb, running crdroid june 15 build, android12.1/12L

#### Issue
i have a specific epub file where the images dont load correctly instead appearing as white  square placeholders only the cover loads correctly. other epubs dont have this problem and it appears correctly in for example read era, or calibre on pc.
#### Steps to reproduce

see attached epub

if it matters all i did was swapped the images using calibre edit book->images->replace illustration with and added a few more sections.

https://drive.google.com/file/d/1JushZFTUVXX1YvmW1VmZgrDP98Wcwdz6/view?usp=sharing

it maybe an issue in that specific epub file but im not expereinced and tought to ask here

### PR DESCRIPTION
**Pull request by poire-z**
Bump base, includes:
- thirdparty/luasec 1.2.0 https://github.com/koreader/koreader-base/pull/1509
- Bump NanoSVG: various small fixes
- Add thirdparty/libwebp and ffi/webp.lua wrapper https://github.com/koreader/koreader-base/pull/1511
- MuPDF: add support for webp images
- bump crengine: add support for webp images
- thirdparty/harfbuzz 5.1.0 https://github.com/koreader/koreader-base/pull/1510

RenderImage:renderImageData(): render webp images with ffi/webp.lua.

Closes #3016. Closes #9240.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9402)

### CODE REVIEW

File: frontend/document/picdocument.lua [hunk 1]
```diff
@@ -63,6 +63,7 @@ function PicDocument:register(registry)
     registry:addProvider("jpg", "image/jpeg", self, 80)
     registry:addProvider("jpeg", "image/jpeg", self, 80)
     registry:addProvider("png", "image/png", self, 80)
+    registry:addProvider("webp", "image/webp", self, 80)
```
Comment (Frenzie) [inline frontend/document/picdocument.lua:66]: Given that you patched MuPDF you can also add this there, presumably with a slightly lower weight. Or does it only work in CBZ?

Comment (poire-z) [reply frontend/document/picdocument.lua:66]: It looks like it doesn't work:
```
08/03/22-16:15:51 INFO  opening file test/webp-sample1.webp
ffi.load: libs/liblept.so.5
ffi.load: libs/libk2pdfopt.so.2
error: cannot recognize version marker
warning: trying to repair broken xref
warning: repairing PDF document
error: no objects found
08/03/22-16:15:52 WARN  cannot open document test/webp-sample1.webp frontend/document/pdfdocument.lua:35: ./ffi/mupdf.lua:78: MuPDF cannot open file.: no objects found (2)
```
The upstream patch probably didn't add the needed stuff to handle webp as the initial document - but only when extracting images from another type of document.
(I'm not interested enough in MuPDF to look at adding that :)

I don't really know in which contexts it works: at least when in CBZ, and when asking MuPDF to render images like we do in frontend (and in that case, it fails with an animated webp - the upstream patch doesn't use the same libwebp objects I ended up using in ffi/webp.lua and crengine - but it should be fine with non-animated webp).

Comment (Frenzie) [reply frontend/document/picdocument.lua:66]: Alright, fair enough.

**Comment by poire-z** [comment]
I feared this would cause issues on Android as added libraries often did - but no, it worked on my phone this morning ! :)

There is a strange thing though, not strictly related to this PR as it happens also with GIF images: on Android (only?), when using File browser with Cover images, the cover/metadata extraction on GIF and WebP images (animated or not) fails (I see with catlog that the subprocess is killed with signal 11 = SIGSEGV).
Anybody with an Android emulator could confirm this also happens there ? (does it also happen on your Android devices ?). And if possible, could anybody get me a gdb stacktrace ?

**Comment by Frenzie** [comment]
I can confirm it doesn't seem to work on Android but I can't provide any more info atm.

**Comment by poire-z** [comment]
Investigated this a bit more.
It crashes when the forked subprocess goes in:
https://github.com/koreader/koreader/blob/9cba1904eecd33c15a71bbd6e63ee122a6aba510/frontend/document/picdocument.lua#L30
https://github.com/koreader/koreader/blob/9cba1904eecd33c15a71bbd6e63ee122a6aba510/frontend/device/android/device.lua#L76
When doing that in the main process, no crash.
So, it might be that the cloned/forked process has not the rights to play with the android stuff ?
Pinging @pazos if that talks to you.

```
08-06 08:03:29.853 23491 23491 W KOReader:  pic20
08-06 08:03:29.853 23491 23491 W KOReader:  pic21
08-06 08:03:29.853 23491 23491 W KOReader:  calling android.isEink
08-06 08:03:29.857 23491 23491 F libc    : Fatal signal 11 (SIGSEGV), code 2, fault addr 0xef30b000 in tid 23491 (Thread-719)
08-06 08:03:29.961   792   792 D clmlib  : Got activities:0x0000000E
08-06 08:03:29.965  1661  2865 W NativeCrashListener: Couldn't find ProcessRecord for pid 23491
08-06 08:03:29.966   792   792 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
08-06 08:03:29.966   792   792 E DEBUG   : AM write failed: Broken pipe
08-06 08:03:29.966   792   792 F DEBUG   : Revision: '0'
08-06 08:03:29.966   792   792 F DEBUG   : ABI: 'arm'
08-06 08:03:29.967   792   792 F DEBUG   : pid: 23491, tid: 23491, name: Thread-719  >>> org.koreader.launcher <<<
08-06 08:03:29.967   792   792 F DEBUG   : signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xef30b000
08-06 08:03:30.004   792   792 F DEBUG   :     r0 00000000  r1 00001000  r2 00000003  r3 eec0d000
08-06 08:03:30.004   792   792 F DEBUG   :     r4 ef30b000  r5 eec0a000  r6 ef408000  r7 f4fdae28
08-06 08:03:30.004   792   792 F DEBUG   :     r8 abf14df8  r9 eec0b000  sl ab43bc78  fp f4fd9910
08-06 08:03:30.004   792   792 F DEBUG   :     ip f4fdae28  sp ef4083f8  lr f4ec5347  pc f4ec534a  cpsr 800f0030
08-06 08:03:30.014   792   792 F DEBUG   :
08-06 08:03:30.014   792   792 F DEBUG   : backtrace:
08-06 08:03:30.014   792   792 F DEBUG   :     #00 pc 0033934a  /system/lib/libart.so (art::Thread::InstallImplicitProtection()+61)
08-06 08:03:30.014   792   792 F DEBUG   :     #01 pc 003394e9  /system/lib/libart.so (art::Thread::InitStackHwm()+188)
08-06 08:03:30.014   792   792 F DEBUG   :     #02 pc 0033970d  /system/lib/libart.so (art::Thread::Init(art::ThreadList*, art::JavaVMExt*, art::JNIEnvExt*)+96)
08-06 08:03:30.014   792   792 F DEBUG   :     #03 pc 003413c1  /system/lib/libart.so (art::Thread::Attach(char const*, bool, _jobject*, bool)+120)
08-06 08:03:30.014   792   792 F DEBUG   :     #04 pc 0031d685  /system/lib/libart.so (art::Runtime::AttachCurrentThread(char const*, bool, _jobject*, bool)+12)
08-06 08:03:30.014   792   792 F DEBUG   :     #05 pc 0024d083  /system/lib/libart.so (art::JII::AttachCurrentThread(_JavaVM*, _JNIEnv**, void*)+142)
08-06 08:03:30.014   792   792 F DEBUG   :     #06 pc 000097f4  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #07 pc 0003b6b3  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #08 pc 00007890  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #09 pc 000490eb  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #10 pc 00007890  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #11 pc 000490eb  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #12 pc 00007890  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so
08-06 08:03:30.014   792   792 F DEBUG   :     #13 pc 00048221  /data/app/org.koreader.launcher-2/lib/arm/libluajit.so (lua_pcall+20)
08-06 08:03:30.014   792   792 F DEBUG   :     #14 pc 00001657  /data/app/org.koreader.launcher-2/lib/arm/libluajit-launcher.so (android_main+298)
08-06 08:03:30.014   792   792 F DEBUG   :     #15 pc 0000232d  /data/app/org.koreader.launcher-2/lib/arm/libluajit-launcher.so
08-06 08:03:30.014   792   792 F DEBUG   :     #16 pc 000418db  /system/lib/libc.so (__pthread_start(void*)+30)
08-06 08:03:30.014   792   792 F DEBUG   :     #17 pc 00019315  /system/lib/libc.so (__start_thread+6)
08-06 08:03:30.014   792   792 F DEBUG   :     #17 pc 00019315  /system/lib/libc.so (__start_thread+6)
```

I remember I wrote this:
https://github.com/koreader/koreader/blob/9cba1904eecd33c15a71bbd6e63ee122a6aba510/frontend/ui/trapper.lua#L460-L468
Dunno if we can/should kill/ghost/make-dummy "android" when forking.
Or add a test in picdocument to not call hasEinkScreen() if we are not the main process (dunno if we can detect that).
(But I don't realy care about this minor issue :)

**Comment by pazos** [comment]
@poire-z: you're right. 

The only place where it is possible to call `android` functions is the main thread, which is attached to the JVM

I guess we can refactor 

```lua
    -- Enforce dithering in PicDocument
    if CanvasContext:hasEinkScreen() then
        if CanvasContext:canHWDither() then
            self.hw_dithering = true
        elseif CanvasContext.fb_bpp == 8 then
            self.sw_dithering = true
        end
    end
```

to something like `Device:canDitherPictures` and just use that in picdocument?

**Comment by poire-z** [comment]
Thanks for confirming the reason.
I guess we can do with just that, which fix the issue for me:
```diff
--- a/frontend/document/canvascontext.lua
+++ b/frontend/document/canvascontext.lua
@@ -55 +55,5 @@ function CanvasContext:init(device)
-    self.hasEinkScreen = device.hasEinkScreen
+    -- This one may be called by a subprocess, and would crash on Android when
+    -- calling android.isEink() which is only allowed from the main thread.
+    local hasEinkScreen = device.hasEinkScreen()
+    self.hasEinkScreen = function() return hasEinkScreen end
+
```

**Comment by Frenzie** [comment]
That looks acceptable.

**Pull request merged**